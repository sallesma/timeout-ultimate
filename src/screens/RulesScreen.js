import React, { useState, useLayoutEffect } from 'react';
import { StyleSheet, View, FlatList, Text, SectionList } from 'react-native';
import { StatusBar } from 'expo-status-bar';
import { List, TextInput } from 'react-native-paper';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import theme from '../utils/theme.js';
import rules from '../../data/rules/rules';
import handSignals from '../../data/rules/handSignals';
import appendix from '../../data/rules/appendix';
import I18n from '../utils/i18n';

import Chapter from '../components/rules/Chapter';
import HandSignal from '../components/rules/HandSignal';
import BottomSheet from '../components/shared/BottomSheet.js';

export default (props) => {
  const [content, setContent] = useState('rules');
  const [isSearchVisible, setIsSearchVisible] = useState(false);
  const [searchText, setSearchText] = useState('');
  const [isSelectorVisible, setIsSelectorVisible] = useState(false);

  useLayoutEffect(() => {
    props.navigation.setOptions({
      headerRight: () => (
        <View style={styles.headerRight}>
          <MaterialCommunityIcons
            name="text-search"
            size={theme.FONT_SIZE_ICON}
            style={styles.headerRightButton}
            onPress={() => setIsSearchVisible(!isSearchVisible)}
          />
          <MaterialCommunityIcons
            name="format-list-bulleted"
            size={theme.FONT_SIZE_ICON}
            style={styles.headerRightButton}
            onPress={() => setIsSelectorVisible(true)}
          />
        </View>
      ),
    });
  });

  const switchContent = (newContent) => {
    setIsSelectorVisible(false);
    setContent(newContent);
  };

  const list = [
    { title: I18n.t('rulesScreen.rules'), onPress: () => switchContent('rules') },
    { title: I18n.t('rulesScreen.handSignals'), onPress: () => switchContent('handSignals') },
    { title: I18n.t('rulesScreen.appendix'), onPress: () => switchContent('appendix') },
  ];

  return (
    <View style={styles.container}>
      {isSearchVisible && (
        <TextInput
          dense
          autoFocus
          mode="outlined"
          value={searchText}
          onChangeText={setSearchText}
          placeholder={I18n.t('rulesScreen.searchPlaceholder')}
          right={<TextInput.Icon icon="close-circle" size={theme.FONT_SIZE_ICON} onPress={() => setSearchText('')} />}
          style={{ marginBottom: 8 }}
        />
      )}
      {content === 'rules' && (
        <FlatList
          keyExtractor={(item, index) => index.toString()}
          data={Object.keys(rules)}
          renderItem={({ item }) => <Chapter title={item} rules={rules[item]} searchText={searchText} />}
          keyboardShouldPersistTaps="handled"
        />
      )}
      {content === 'handSignals' && (
        <FlatList
          keyExtractor={(item, index) => index.toString()}
          data={handSignals}
          renderItem={({ item }) => <HandSignal item={item} searchText={searchText} />}
          keyboardShouldPersistTaps="handled"
        />
      )}
      {content === 'appendix' && (
        <SectionList
          keyExtractor={(item, index) => index.toString()}
          sections={Object.keys(appendix).map((appendixSection) => ({
            title: appendixSection,
            data: Object.keys(appendix[appendixSection]),
          }))}
          renderSectionHeader={({ section }) => (
            <View style={styles.appendixSection}>
              <Text style={styles.appendixSectionText}>{section.title}</Text>
            </View>
          )}
          renderItem={({ item, section }) => (
            <Chapter title={item} rules={appendix[section.title][item]} searchText={searchText} />
          )}
        />
      )}
      <BottomSheet onClose={() => setIsSelectorVisible(false)} isVisible={isSelectorVisible}>
        {list.map((listItem, i) => (
          <List.Item key={i} title={listItem.title} onPress={listItem.onPress} />
        ))}
      </BottomSheet>
      <StatusBar style="auto" />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
  },
  headerRight: {
    flexDirection: 'row',
    marginRight: 16,
  },
  headerRightButton: {
    color: 'white',
    padding: 8,
  },
  appendixSection: {
    backgroundColor: theme.MAIN_COLOR_LIGHT,
  },
  appendixSectionText: {
    fontSize: theme.FONT_SIZE_L,
    fontWeight: 'bold',
    color: theme.MAIN_COLOR,
  },
});
