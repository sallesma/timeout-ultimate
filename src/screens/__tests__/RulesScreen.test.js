import React from 'react';
import { render } from '@testing-library/react-native';

import RulesScreen from '../RulesScreen';

describe('<RulesScreen />', () => {
  it('renders correctly', async () => {
    const navigation = { setOptions: jest.fn() };

    const { toJSON } = render(<RulesScreen navigation={navigation} />);

    expect(toJSON()).toMatchSnapshot();
  });
});
