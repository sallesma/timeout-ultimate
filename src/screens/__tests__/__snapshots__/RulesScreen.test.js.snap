// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RulesScreen /> renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 16,
    }
  }
>
  <RCTScrollView
    data={
      [
        "1.L’Esprit du Jeu",
        "2.Aire de Jeu",
        "3.Équipement",
        "4.Point et Jeu",
        "5.Équipes",
        "6.Commencer un Match",
        "7.Le « Pull » (Engagement)",
        "8.Etat du Jeu",
        "9.« Stall Count » (Le Compte)",
        "10.Le « Check »",
        "11.« Out » (Hors-limites)",
        "12.Réceptionneur et Positionnement",
        "13.« Turnovers »",
        "14.Marquer un point",
        "15.Appeler les Fautes, Infractions et Violations",
        "16.Continuité après un appel",
        "17.Fautes",
        "18.Infractions et Violations",
        "19.Interruptions de sécurité",
        "20.Temps-Morts (« Time-Outs »)",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              1. L’Esprit du Jeu
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              2. Aire de Jeu
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              3. Équipement
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              4. Point et Jeu
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              5. Équipes
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              6. Commencer un Match
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              7. Le « Pull » (Engagement)
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              8. Etat du Jeu
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              9. « Stall Count » (Le Compte)
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#2a9d9d",
                "borderBottomWidth": 0.5,
                "flexDirection": "row",
                "marginTop": 8,
                "opacity": 1,
                "paddingVertical": 8,
              }
            }
          >
            <Text />
            <Text
              style={
                {
                  "fontSize": 20,
                  "marginLeft": 8,
                }
              }
            >
              10. Le « Check »
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
